### PR TITLE
Major refactor separating tracing, writing, and formatting

### PR DESCRIPTION
**Pull request by alexmojaki**
This separates the concerns of tracing (gathering data), formatting the data, and writing the formatted data. The code is much more organised and easier to read, modify, and test, and the library is easier to externally customise and extend.

The next stage is to remove the use of regexes from tests. Instead, test entries will be compared with the unformatted data directly.

After that, I'd like to make my own separate library which uses pysnooper at its core while adding additional functionality through dependencies. For example, I'd like to have output colouring including syntax highlighting using pygments and colorama.

In the long term, this will also enable a radically different infrastructure such as writing to a database and then viewing the logs in a web UI.

And then, I'd like to understand how to git.

### CODE REVIEW

**Comment by cool-RR** [comment]
God damn. This might take me a while.

File: pysnooper/formatting.py [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+            **entry.__dict__
+        )
+
+    def format_ReturnValueEntry(self, entry, _event):
+        return u'{description} {value_repr}'.format(
+            description=u'Return value:'.ljust(self.now_length, '.'),
+            value_repr=utils.get_shortish_repr(entry.value),
+        )
+
+
```
Comment (alexmojaki) [inline pysnooper/formatting.py]: Below this line in this file things are only copied, not changed.

**Comment by alexmojaki** [comment]
OK, I've vastly simplified the tests, but now they're not covering everything (particularly formatting) so I will be adding some more soon, though probably not tonight.

File: pysnooper/tracer.py [hunk 2]
```diff
@@ -245,17 +151,14 @@ def trace(self, frame, event, arg):
                 return None
             else:
                 _frame_candidate = frame
-                for i in range(1, self.depth):
+                for depth in range(1, self.depth):
```
[the remark this answers is not in the document]
Comment (alexmojaki) [reply pysnooper/tracer.py]: It's a bit weird, the loop's purpose is to assign `depth` to be used further down to make an `Event`. `indent` is the formatter's job now.

**Comment by cool-RR** [comment]
Rebasing would be preferable to merging.

**Comment by cool-RR** [comment]
If you'll want help learning to rebase, you can send me your repo zipped to ram@rachum.com and I'll show you the list of commands I used to fix things up.

**Comment by alexmojaki** [comment]
Closing in favour of #103

**Pull request closed without merging**